core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -80,8 +86,59 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
   //periodically emit metrics (don't need to do this for each message!)
   context.system.scheduler.schedule(30.seconds, 10.seconds, self, EmitMetrics)
 
+  // Key is `kind,memory,time`, value is the number of cold Start in minute
+  var coldStartCount = immutable.Map.empty[String, Int]
+
   backfillPrewarms(true)
 
+  // check periodically every 1 minute, delete prewarmed container if unused for sometime
+  context.system.scheduler.schedule(10.seconds, 1.minute, self, DeleteUnusedPrewarmedContainer)
+
+  // check periodically for the cold start and create some increment containers automatically if activation >= threshold
+  context.system.scheduler.schedule(1.minute, 1.minute, self, SupplementPrewarmedContainer)
+
+  def deleteUnusedPrewarmedContainer(): Unit = {
+    prewarmConfig.foreach { config =>
+      val kind = config.exec.kind
+      val memory = config.memoryLimit
+      val ttlSeconds = config.ttl.toSeconds
+      val containers = prewarmedPool.filter { warmInfo =>
+        warmInfo match {
+          case (_, PreWarmedData(_, `kind`, `memory`, _)) => true
+          case _                                          => false
+        }
+      }
+      for ((container, data) <- containers) {
+        if (JDuration.between(data.lastUsed, Instant.now).compareTo(JDuration.ofSeconds(ttlSeconds)) > 0) {
+          // Don't recover a new one under this situation
+          container ! RemovePreWarmedContainer
```

Comment:
Can you just use Remove message here?

Reply:
If use `Remove` message to remove unused prewarmed container, this lead to another new prewarmed created after delete the unused prewarmed container.

Logic as below
 firstly, codes will be executed here: https://github.com/apache/openwhisk/blob/master/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerProxy.scala#L376, then after executed `context.parent ! ContainerRemoved` in `def destroyContainer`, codes will be executed here: https://github.com/apache/openwhisk/blob/master/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala#L272

So i used another message `RemovePreWarmedContainer` to distinguish them, and has another benefit, after deleted the prewarmed container, can print the prewarmed container's detail info via `PreWarmedContainerRemoved(prewarmedData: PreWarmedData)`'s `prewarmedData` field

Reply:
I see. In that case, I suggest changing `case object ContainerRemoved` to `case class ContainerRemoved(replacePrewarm:Booleam)` so that we can limit the duplicated logic in `ContainerProxy` - would that work?

Reply:
Yes, can work, updated accordingly